pkg/idp/provider/azure/azure.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/giantswarm/microerror"
+	azauth "github.com/microsoft/kiota-authentication-azure-go"
+	msgraphsdk "github.com/microsoftgraph/msgraph-sdk-go"
+	"github.com/microsoftgraph/msgraph-sdk-go/models"
+)
+
+type Azure struct {
+	Name     string
+	Client   *msgraphsdk.GraphServiceClient
+	Owner    string
+	TenantID string
+	Type     string
+}
+
+func New(p provider.ProviderCredential) (*Azure, error) {
+	var tenantID, clientID, clientSecret string
+	{
+		if p.Name == "" {
+			return nil, microerror.Maskf(invalidConfigError, "Credential name must not be empty.")
+		}
+		if p.Owner == "" {
+			return nil, microerror.Maskf(invalidConfigError, "Credential owner must not be empty.")
+		}
+		if tenantID = p.Credentials[TenantIDKey]; tenantID == "" {
+			return nil, microerror.Maskf(invalidConfigError, "%s must not be empty.", TenantIDKey)
+		}
+		if clientID = p.Credentials[ClientIDKey]; clientID == "" {
+			return nil, microerror.Maskf(invalidConfigError, "%s must not be empty.", ClientIDKey)
+		}
+		if clientSecret = p.Credentials[ClientSecretKey]; clientSecret == "" {
+			return nil, microerror.Maskf(invalidConfigError, "%s must not be empty.", ClientSecretKey)
+		}
+	}
+	cred, err := azidentity.NewClientSecretCredential(tenantID, clientID, clientSecret, nil)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	auth, err := azauth.NewAzureIdentityAuthenticationProviderWithScopes(cred, ProviderScope())
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	adapter, err := msgraphsdk.NewGraphRequestAdapter(auth)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	client := msgraphsdk.NewGraphServiceClient(adapter)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	return &Azure{
+		Name:     fmt.Sprintf("%s-%s", p.Owner, p.Name),
+		Type:     ProviderConnectorType,
+		Client:   client,
+		Owner:    p.Owner,
+		TenantID: tenantID,
+	}, nil
+}
+
+func (a *Azure) GetName() string {
+	return a.Name
+}
+
+func (a *Azure) GetType() string {
+	return a.Type
+}
+
+func (a *Azure) GetOwner() string {
+	return a.Owner
+}
+
+func (a *Azure) CreateApp(config provider.AppConfig, ctx context.Context) (dex.Connector, error) {
+	createdApp, err := a.Client.Applications().Post(ctx, getAppCreateRequestBody(config), nil)
+	if err != nil {
+		return dex.Connector{}, microerror.Maskf(requestFailedError, printOdataError(err))
+	}
+	id := createdApp.GetId()
+	if id == nil {
+		return dex.Connector{}, microerror.Maskf(notFoundError, "Could not find ID of app %s.", config.Name)
+	}
+	createdSecret, err := a.Client.ApplicationsById(*id).AddPassword().Post(ctx, getSecretCreateRequestBody(config), nil)
+	if err != nil {
+		return dex.Connector{}, microerror.Maskf(requestFailedError, printOdataError(err))
+	}
+	parentApp, err := a.GetApp(DefaultName)
+	if err != nil {
+		return dex.Connector{}, microerror.Mask(err)
+	}
+	_, err = a.Client.ApplicationsById(*id).Patch(ctx, getPermissionCreateRequestBody(parentApp), nil)
```

Comment:
Adding the permissions based on permissions set on a parent app.
The reason we do it that way is that querying permissions by scope (human readable) is not possible in the microsoft graph API and hardcoding the IDs seems risky